### PR TITLE
The design draft of adding gateway to mizar subnets

### PR DESCRIPTION
**Pull request by jshaofuturewei**

### CODE REVIEW

File: docs/fornax-design/mizar_gateway_design [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+## Data Plane
+A new type gateway droplet is added. Gateway bouncers will be deployed in that droplet and foward requests to the gateway
+
+### RPC service 
+1. A new service to update subnet by adding the gateway and the external fields
+2. A new service to update bouncers by assigning gateway ip
+
+### Transit XDP
+1. A bouncer metadata contain divider list
```
Comment (pdgetrf) [inline docs/fornax-design/mizar_gateway_design:54]: why is bouncer involved in gateway?

File: docs/fornax-design/mizar_gateway_design [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+```
+
+## Control Plane
+For mizar control planes, we add the two fields above to subnet objects. The default value of gateway is empty and the one of external is false. If all the requests happen within the clusters, the two fields do not invovle at all.
+
+### Create a gateway droplet
+1. Deploy gateway endpoints
+2. Get gateway endpoint droplet ip 
+3. Add the gateway droplet ip to subnet
+4. The change triggers subnet workflows to delete the droplet and treat it as gateway droplet
```
Comment (pdgetrf) [inline docs/fornax-design/mizar_gateway_design:30]: this is not consistent with our POC method. what's the reason of the divergence?

File: docs/fornax-design/mizar_gateway_design [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+### Create an external subnet
+1. Deploy an external cluster
+2. Deploy a subnet and its gateway in the external cluster
+3. Create an external subnet with the gateway ip in a local cluster
+4. THe control plane does not trigger any other creation since the subnet is external
+
+### Ping external endpints in a local cluster
+1. Suppose the local subnet is 20.0.0.0/24 while the external subnet is 30.0.0.0/24.
```
Comment (pdgetrf) [inline docs/fornax-design/mizar_gateway_design:39]: subnets need to be in the same VPC. can 20.0.0.0/24 and 30.0.0.0/24 be from the same VPC？

File: docs/fornax-design/mizar_gateway_design [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+### Ping external endpints in a local cluster
+1. Suppose the local subnet is 20.0.0.0/24 while the external subnet is 30.0.0.0/24. 
+2. A request is created in 20.0.0.10 to request the service at 30.0.0.10.
+3. The request will treat as external request because of the target ip 30.0.0.10.
+4. Given the source ip is 20.0.0.10, the transit xdp finds its local gateway and redirect the request to the gateway
+5. The gateway accepts its request and found its external subnet gateway
+6. The gateway redirects the request to the external gateaway
+
+## Data Plane
+A new type gateway droplet is added. Gateway bouncers will be deployed in that droplet and foward requests to the gateway
```
Comment (pdgetrf) [inline docs/fornax-design/mizar_gateway_design:47]: I think we've discussed this a few times. Why is bouncer involved in the gateway design at all?

File: docs/fornax-design/mizar_gateway_design [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+### RPC service 
+1. A new service to update subnet by adding the gateway and the external fields
+2. A new service to update bouncers by assigning gateway ip
```
Comment (pdgetrf) [inline docs/fornax-design/mizar_gateway_design:51]: same question as before. why is bouncer involved?

File: docs/fornax-design/mizar_gateway_design (hunk 1 above)
Comment (jshaofuturewei) [reply docs/fornax-design/mizar_gateway_design:54]: I am thinking to keep the origin bouncer and divider design without updating too many codes.  I am testing it to see if it works. I will roll back it to divider.

File: docs/fornax-design/mizar_gateway_design (hunk 2 above)
Comment (jshaofuturewei) [reply docs/fornax-design/mizar_gateway_design:30]: POC is to have a predefined host as gateway. I just make it configurable. After we deploy a gateway and know its ip, we will update the subnet by adding the gateway ip. That ip will be removed from droplet set as POC does

File: docs/fornax-design/mizar_gateway_design (hunk 3 above)
Comment (jshaofuturewei) [reply docs/fornax-design/mizar_gateway_design:39]: It is very confusing, I will correct it.  They can be the same vpc of the 16.0.0.0/4.  All the subnets can the in the same vpc, such as 0.0.0.0/0?

File: docs/fornax-design/mizar_gateway_design (hunk 4 above)
Comment (jshaofuturewei) [reply docs/fornax-design/mizar_gateway_design:47]: I will change it to divider. We can discuss later about the possible code changes

**Pull request closed without merging**